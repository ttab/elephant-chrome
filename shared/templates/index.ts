import type { Block } from '@ttab/elephant-api/newsdoc'

export { articleDocumentTemplate as article } from './articleDocumentTemplate.js'
export { flashDocumentTemplate as flash } from './flashDocumentTemplate.js'
export { factboxDocumentTemplate as factbox } from './factboxDocumentTemplate.js'
export { assignmentPlanningTemplate as assignment } from './assignmentPlanningTemplate.js'
export { planningDocumentTemplate as planning } from './planningDocumentTemplate.js'
export { eventDocumentTemplate as event } from './eventDocumentTemplate.js'
export { editorialInfoDocumentTemplate as editorialInfo } from './editorialInfoDocumentTemplate.js'

export interface TemplatePayload {
  title?: string
  meta?: {
    'tt/slugline'?: Block[]
    'core/newsvalue'?: Block[]
    'core/planning-item'?: Block[]
    'core/description'?: Block[]
    'core/event'?: Block[]
  }
  links?: {
    'core/section'?: Block[]
    'core/story'?: Block[]
    'tt/wire'?: Block[]
    'core/event'?: Block[]
  }
}
