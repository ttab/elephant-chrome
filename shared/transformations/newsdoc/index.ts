import { transformText, revertText, transformUnorderedList } from './core/index.js'
import { transformVisual, revertVisual } from './tt/visual.js'
import type { TBElement } from '@ttab/textbit'
import type { Block } from '@ttab/elephant-api/newsdoc'
import { revertFactbox, transformFactbox } from './core/factbox.js'
import { revertTable, transformTable } from './core/table.js'
import { revertTvListing, transformTvListing } from './tt/tvListing.js'
import { revertPrintArticle, transformPrintArticle } from './tt/printArticle.js'

/**
 * Convert a NewsDoc block array to slate TBElement array
 */
export function newsDocToSlate(content: Block[]): TBElement[] {
  if (content !== undefined && Array.isArray(content)) {
    return content.map((element: Block): TBElement => {
      switch (element.type) {
        case 'core/text':
        case 'tt/print-text':
          return transformText(element)
        case 'tt/visual':
          return transformVisual(element)
        case 'core/factbox':
          return transformFactbox(element)
        case 'core/unordered-list':
          return transformUnorderedList(element)
        case 'core/table':
          return transformTable(element)
        case 'tt/tv-listing':
          return transformTvListing(element)
        case 'tt/print-article':
          return transformPrintArticle(element)
        default:
          throw new Error(`Element not implemented: ${element.type}`)
      }
    })
  }
  throw new Error('No document to transform')
}

/**
 * Convert a slate TBElement array to a NewsDoc block array
 */
export function slateToNewsDoc(elements: TBElement[]): Block[] | undefined {
  if (!Array.isArray(elements)) {
    return []
  }

  return elements.map((element: TBElement) => {
    switch (element.type) {
      case 'core/text':
        return revertText(element)

      case 'tt/visual':
      case 'core/image':
        return revertVisual(element)
      case 'core/factbox':
        return revertFactbox(element)
      case 'core/table':
        return revertTable(element)
      case 'tt/tv-listing':
        return revertTvListing(element)
      case 'tt/print-article':
        return revertPrintArticle(element)
      default:
        throw new Error(`Element not implemented: ${element.type}`)
    }
  })
}
