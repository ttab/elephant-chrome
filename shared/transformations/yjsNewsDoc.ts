import * as Y from 'yjs'
import { type Document } from '@hocuspocus/server'

import type {
  EleBlock,
  EleDocumentResponse
} from '@/shared/types/index.js'
import { toYMap } from './toYMap.js'
import { slateNodesToInsertDelta, yTextToSlateElement } from '@slate-yjs/core'
import createHash from '../createHash.js'
import { type TBElement } from '@ttab/textbit'
import { transformYXmlTextNodes } from './lib/transformYXmlTextNodes.js'

/**
 * Convert a grouped YDocumentResponse a yjs structure and add it to the provided Y.Doc
 */
export function toYjsNewsDoc(eleDoc: EleDocumentResponse, yDoc: Document | Y.Doc): void {
  const yMap = yDoc.getMap('ele')
  const { document, version } = eleDoc

  if (!document) {
    throw new Error('YDocumentResponse contains no document')
  }

  const { meta, links, content, ...properties } = document
  yMap.set('meta', toYMap(meta))
  yMap.set('links', toYMap(links))
  yMap.set('root', toYMap(properties, new Y.Map()))

  // Slate text to yjs
  const yContent = new Y.XmlText()
  yContent.applyDelta(
    slateNodesToInsertDelta(content)
  )

  yMap.set('content', yContent)

  // Set version
  const yVersion = yDoc.getMap('version')
  yVersion?.set('version', version)

  // Create original hash based on eleDocument content and store in yjs structure
  const originalHash = createHash(JSON.stringify(eleDoc.document))
  const yOriginalHash = yDoc.getMap('hash')
  yOriginalHash?.set('hash', originalHash)
}


/**
 * Convert a yjs structure to a grouped YDocumentResponse
 */
export function fromYjsNewsDoc(yDoc: Y.Doc): {
  documentResponse: EleDocumentResponse
  updatedHash: number | undefined
  originalHash: number
} {
  const yMap = yDoc.getMap('ele')

  const root = yMap.get('root') as Y.Map<unknown>
  const { uuid, type, uri, url, title, language } = root.toJSON() as Record<string, string>

  const meta = transformYXmlTextNodes(yMap.get('meta') as Y.Map<unknown>) as Record<string, EleBlock[]>

  const links = (yMap.get('links') as Y.Map<unknown>).toJSON() || {}

  const yContent = yMap.get('content') as Y.XmlText
  const content = yContent.length ? yTextToSlateElement(yContent).children : []

  const responseDocument = {
    version: yDoc.getMap('version').get('version') as string,
    isMetaDocument: false,
    mainDocument: '',
    document: {
      uuid,
      type,
      uri,
      url,
      title,
      content: content as TBElement[],
      meta,
      links,
      language
    }
  }

  const currentHash = createHash(JSON.stringify(responseDocument.document))
  const originalHash = (yDoc.getMap('hash')?.get('hash') || 0) as number

  // Return response, original hash from yjs structure and newly calculated hash
  return {
    documentResponse: responseDocument,
    updatedHash: (currentHash !== originalHash) ? currentHash : undefined,
    originalHash
  }
}
