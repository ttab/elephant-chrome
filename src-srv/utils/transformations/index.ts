export {
  slateToNewsDoc,
  newsDocToSlate
} from './newsdoc/index.js'
