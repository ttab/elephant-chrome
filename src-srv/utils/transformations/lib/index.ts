export { toYMap } from './toYMap.js'
export { toSlateYXmlText } from './toSlateYXmlText.js'
export { textToNewsDoc } from './textToNewsdoc.js'
