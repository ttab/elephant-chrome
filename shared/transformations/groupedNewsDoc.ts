import { type GetDocumentResponse } from '@ttab/elephant-api/repository'
import { Block, type Document } from '@ttab/elephant-api/newsdoc'

import type {
  EleDocumentResponse,
  EleDocument,
  EleBlockGroup
} from '@/shared/types/index.js'

import { newsDocToSlate, slateToNewsDoc } from './newsdoc/index.js'


/**
 *  Convert repository format NewsDoc to grouped YDocument format
 */
export const toGroupedNewsDoc = (payload: GetDocumentResponse): EleDocumentResponse => {
  // Create clone of document so not to change original document/cause sideffects
  const { document, version, isMetaDocument, mainDocument } = structuredClone(payload)

  if (!document) {
    throw new Error('GetDocumentResponse contains no document')
  }

  if (document.type === 'core/planning-item') {
    assertPlanningHasNecessaryProperties(document)
  }

  const yDocument: EleDocument = {
    ...document,
    content: newsDocToSlate(document.content, document.type === 'tt/print-article'),
    meta: group(document.meta || [], 'type'),
    links: group(document.links || [], 'type')
  }

  return {
    version: version.toString(),
    isMetaDocument,
    mainDocument,
    document: yDocument
  }
}


/**
 *  Convert grouped YDocument format to the repository format NewsDoc
 */
export const fromGroupedNewsDoc = (payload: EleDocumentResponse): { document: Document } & Omit<GetDocumentResponse, 'document'> => {
  const { document, version, isMetaDocument, mainDocument } = payload

  if (!document) {
    throw new Error('YDocumentResponse contains no document')
  }

  const newsDocument = {
    ...document,
    content: slateToNewsDoc(document.content) || [],
    meta: ungroup(document.meta),
    links: ungroup(document.links)
  }

  if (document.type === 'core/planning-item') {
    assertPlanningHasNoEmptyProperties(newsDocument)
  }

  return {
    version: BigInt(version),
    document: newsDocument,
    isMetaDocument,
    mainDocument
  }
}


/**
 * Group Blocks with same group key in arrays
 *
 * @param Block[]
 * @param groupKey
 * @returns YBlockGroup
 */
export function group(objects: Block[], groupKey: keyof Block): EleBlockGroup {
  const groupedObjects: EleBlockGroup = {}

  objects.forEach((object) => {
    let key = object[groupKey] as string | undefined

    // Fallback to underscore if no groupKey is found
    if (!key) {
      key = '_'
    }

    if (!groupedObjects[key]) {
      groupedObjects[key] = []
    }

    const newObj = { ...object }

    Object.entries(newObj).forEach(([key, value]) => {
      if (Array.isArray(value)) {
        // @ts-expect-error Can't get this ts error to go away
        newObj[key] = group(value, groupKey)
      }
    })

    // @ts-expect-error Can't get this ts error to go away
    groupedObjects[key].push(newObj)
  })

  return groupedObjects
}


/**
 * Reverts the operation of the group function
 *
 * @param YBlockGroup
 * @returns Block[]
 */
export function ungroup(obj: EleBlockGroup): Block[] {
  const result: Block[] = []

  Object.keys(obj).forEach((key) => {
    if (Array.isArray(obj[key])) {
      obj[key].forEach((item) => {
        // Ignore __inProgress items
        if (!item.__inProgress) {
          const newObj = Block.create({
            ...item,
            meta: ungroup(item.meta as unknown as EleBlockGroup || {}),
            links: ungroup(item.links as unknown as EleBlockGroup || {}),
            content: ungroup(item.content as unknown as EleBlockGroup || {})
          })

          result.push(newObj as unknown as Block)
        }
      })
    }
  })

  return result
}

/**
 *  Assert that we don't leave empty sluglines or descriptions on planning documents
 */
function assertPlanningHasNoEmptyProperties(obj: Document | Block): void {
  obj.meta = obj.meta.filter((block) => {
    if (block.type === 'core/description' && !block.data.text) {
      return false
    } else if (block.type === 'tt/slugline' && !block.value) {
      return false
    }

    return true
  })

  obj.meta.forEach((block) => {
    if (block.type === 'core/assignment') {
      assertPlanningHasNoEmptyProperties(block)
    }
  })
}


/**
 *  Assert that slugline and both internal and public descriptions exist on planning documents
 */
function assertPlanningHasNecessaryProperties(obj: Document | Block): void {
  const internalDesc = obj.meta.find((item) => item.type === 'core/description' && item.role === 'internal')
  const publicDesc = obj.meta.find((item) => item.type === 'core/description' && item.role === 'public')
  const slugline = obj.meta.find((item) => item.type === 'tt/slugline')

  if (!slugline) {
    obj.meta.push(Block.create({
      type: 'tt/slugline'
    }))
  }

  if (!internalDesc) {
    obj.meta.push(Block.create({
      type: 'core/description',
      role: 'internal',
      data: { text: '' }
    }))
  }

  if (!publicDesc) {
    obj.meta.push(Block.create({
      type: 'core/description',
      role: 'public',
      data: { text: '' }
    }))
  }

  obj.meta.forEach((block) => {
    if (block.type === 'core/assignment') {
      assertPlanningHasNecessaryProperties(block)
    }
  })
}
