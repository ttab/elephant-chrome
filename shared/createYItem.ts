import * as Y from 'yjs'
import { Block, type Document } from '@ttab/elephant-api/newsdoc'
import type { Wire } from '@/shared/schemas/wire.js'
import { assignmentPlanningTemplate } from '@/shared/templates/assignmentPlanningTemplate.js'
import type { DeliverableType } from '@/shared/templates/lib/getDeliverableType.js'
import { group, toGroupedNewsDoc } from '@/shared/transformations/groupedNewsDoc.js'
import { toYMap } from '@/shared/transformations/toYMap.js'
import { toYjsNewsDoc } from '@/shared/transformations/yjsNewsDoc.js'
import type { IDBAuthor } from '../src/datastore/types.js'

/**
* General function to create a new document as Y.Doc from a template
* @returns [string, Y.Doc]
*/
interface CreateDocumentParams<T> {
  template: (documentId: string, payload?: T) => Document
  inProgress?: boolean
  payload?: T
  createdDocumentIdRef?: React.MutableRefObject<string | undefined>
  documentId?: string
}

export function createDocument<T>({
  template,
  inProgress,
  payload,
  createdDocumentIdRef,
  documentId
}: CreateDocumentParams<T>): [string, Y.Doc] {
  // Use provided documentId or generate a new one
  const docId = documentId || crypto.randomUUID()

  // Set generated documentId to ref so that it can be
  // accessed from creating component
  if (createdDocumentIdRef) {
    createdDocumentIdRef.current = docId
  }

  const yDoc = new Y.Doc()

  toYjsNewsDoc(
    toGroupedNewsDoc({
      version: 0n,
      isMetaDocument: false,
      mainDocument: '',
      document: template(docId, payload)
    }),
    yDoc
  )

  if (inProgress) {
    yDoc.getMap('__meta').set('isInProgress', true)
  }

  return [docId, yDoc]
}

/**
* Specific function to create a new assignment as Y.Map from a template
* and append it to the meta yMap
* @returns void
*/
export function appendAssignment({
  assignee,
  document,
  type,
  inProgress,
  slugLine,
  title,
  wire,
  assignmentData
}: {
  document: Y.Doc
  assignee?: IDBAuthor | null | undefined
  type: 'text' | 'flash' | 'graphic' | 'picture' | 'video' | 'picture/video'
  inProgress?: { sub: string }
  slugLine?: string
  title?: string
  wire?: Wire
  assignmentData?: Block['data']
}): [number, Y.Map<unknown>] {
  const meta = document.getMap('ele').get('meta') as Y.Map<unknown>

  // Get slugline from planning
  const slugLineArray = meta?.get('tt/slugline') as Y.Array<unknown>
  const slugLineYXml = slugLineArray?.get(0) as Y.Map<unknown>
  const slugLineFromPlanning = (slugLineYXml?.get('value') as Y.XmlText)?.toString() as string || undefined

  // Check if 'core/assignment' exists
  if (!meta.has('core/assignment')) {
    meta.set('core/assignment', new Y.Array())
  }

  // Get existing assignments
  const yAssignments = meta.get('core/assignment') as Y.Array<unknown>

  const yMeta = meta.get('core/planning-item') as Y.Array<Y.Map<unknown>>
  const planningDate = (yMeta?.get(0)?.get('data') as Y.Map<unknown>)
    ?.get('start_date') as string

  // Create new assignment from template
  const assignment = assignmentPlanningTemplate({
    assignee,
    assignmentType: type,
    planningDate,
    slugLine: type !== 'flash' ? slugLine || slugLineFromPlanning : undefined,
    title: title,
    wire,
    assignmentData
  })

  // Append __inProgress if needed
  if (inProgress) {
    // @ts-expect-error We need to override Block to add this property
    assignment.__inProgress = inProgress
  }

  // Group assignment
  const [groupedAssignment] = group([assignment], 'type')['core/assignment']

  // Convert to YMap
  const yAssignment = toYMap(
    groupedAssignment as unknown as Record<string, unknown>,
    new Y.Map()
  )

  // Push to existing assignments
  yAssignments.push([yAssignment])

  return [
    yAssignments.length - 1,
    yAssignment
  ]
}

/**
* Specific function to create a new article in Y.Doc as Y.Map from a template
*/
export function appendDocumentToAssignment({ document, id, index, slug, type }: {
  document: Y.Doc
  id: string
  index: number
  slug?: string
  type: DeliverableType
}): void {
  // Get meta yMap
  const meta = document.getMap('ele').get('meta') as Y.Map<unknown>

  // Get assignmentLinks
  const assignmentLinks = ((meta
    .get('core/assignment') as Y.Array<unknown>)
    .get(index) as Y.Map<unknown>)
    .get('links') as Y.Map<unknown>

  // Check if deliverableType exists
  if (!assignmentLinks.has(`core/${type}`)) {
    assignmentLinks.set(`core/${type}`, new Y.Array())
  }
  // Get existing articles
  const yDeliverables = assignmentLinks.get(`core/${type}`) as Y.Array<unknown>

  // Create new deliverable from template
  const deliverable = Block.create({
    type: `core/${type}`,
    uuid: id,
    rel: 'deliverable',
    title: slug
  })

  // Group deliverable
  const [groupedDeliverable] = group([deliverable], 'type')[`core/${type}`]

  // Convert to YMap
  const yDeliverable = toYMap(
    groupedDeliverable as unknown as Record<string, unknown>,
    new Y.Map()
  )
  // Push to existing articles
  yDeliverables.push([yDeliverable])
}
