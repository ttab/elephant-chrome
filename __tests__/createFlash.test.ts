import { createFlash } from '@/views/Flash/lib/createFlash'
import { createDocument } from '@/lib/createYItem'
import * as Templates from '@/defaults/templates'
import type { Session } from 'next-auth'
import { vi } from 'vitest'
import type { HocuspocusProvider } from '@hocuspocus/provider'
import { fromYjsNewsDoc } from '@/shared/transformations/yjsNewsDoc'
import { fromGroupedNewsDoc } from '@/shared/transformations/groupedNewsDoc'
import type { Document } from '@ttab/elephant-api/newsdoc'
import { Block } from '@ttab/elephant-api/newsdoc'

const ASSIGNMENT_TITLE = 'Test flash'
const PLANNING_TITLE = 'Test planning'
const NEWSVALUE = '4'
const SECTION = 'Inrikes'
const session = { accessToken: 'abc123' } as Session

describe('createFlash', () => {
  const mockRandomUUID = vi.spyOn(crypto, 'randomUUID')
  mockRandomUUID.mockImplementation(() => `uuid-${mockRandomUUID.mock.calls.length}`)

  const mockDate = new Date(Date.UTC(2025, 1, 17))
  vi.spyOn(global, 'Date').mockImplementation(() => mockDate)

  const setupDocuments = (isExistingPlanning: boolean) => {
    const planningPayload = isExistingPlanning
      ? {
          title: PLANNING_TITLE,
          meta: { 'core/newsvalue': [Block.create({ type: 'core/newsvalue', value: NEWSVALUE })] },
          links: { 'core/section': [Block.create({ type: 'core/section', title: SECTION })] }
        }
      : {
          meta: { 'core/newsvalue': [Block.create({ type: 'core/newsvalue', value: NEWSVALUE })] }
        }

    const planning = createDocument({ template: Templates.planning, payload: planningPayload })
    const flash = createDocument({
      template: Templates.flash,
      payload: {
        title: ASSIGNMENT_TITLE,
        section: { title: SECTION, uuid: '956636ed-2687-4bc3-a45a-7e09d98c6eeb' }
      }
    })

    return { planning, flash }
  }

  const runCreateFlash = (isExistingPlanning: boolean) => {
    const { planning, flash } = setupDocuments(isExistingPlanning)
    const sendStateless = vi.fn()
    const provider = { document: flash[1], sendStateless } as unknown as HocuspocusProvider

    const result = createFlash({
      provider,
      status: 'authenticated',
      session,
      planning: { document: planning[1], id: planning[0] },
      hasSelectedPlanning: isExistingPlanning,
      timeZone: 'Europe/Stockholm',
      documentStatus: undefined
    })

    if (!result || !result.planning || !result.flash) {
      throw new Error('No planning or flash document created')
    }

    const planningYjs = fromYjsNewsDoc(result.planning).documentResponse
    const flashYjs = fromYjsNewsDoc(result.flash).documentResponse
    return { flash: fromGroupedNewsDoc(flashYjs), planning: fromGroupedNewsDoc(planningYjs), sendStateless }
  }

  describe('New planning', () => {
    let flash: { document: Document }
    let planning: { document: Document }
    let sendStateless: ReturnType<typeof vi.fn>

    beforeAll(() => {
      ({ flash, planning, sendStateless } = runCreateFlash(false))
    })

    describe('Snapshots', () => {
      it('should match planning snapshot', () => {
        expect(planning).toMatchSnapshot()
      })

      it('should match flash snapshot', () => {
        expect(flash).toMatchSnapshot()
      })
    })

    describe('Section', () => {
      it('should only have one section block in each document', () => {
        const flashSection = flash.document.links.filter((l: Block) => l.type === 'core/section')
        const planningSection = planning.document.links.filter((l: Block) => l.type === 'core/section')

        expect(flashSection).toHaveLength(1)
        expect(planningSection).toHaveLength(1)
      })

      it('should have the same section in both planning and article', () => {
        const flashSection = flash.document.links.filter((l: Block) => l.type === 'core/section')
        const planningSection = planning.document.links.filter((l: Block) => l.type === 'core/section')

        expect(flashSection[0].title).toEqual(SECTION)
        expect(flashSection).toEqual(planningSection)
      })
    })

    describe('Slugline', () => {
      it('should not have a slugline block, neither in flash or planning', () => {
        const flashSlugline = flash.document.meta.filter((m: Block) => m.type === 'tt/slugline')
        const planningSlugline = planning.document.meta.filter((m: Block) => m.type === 'tt/slugline')

        expect(flashSlugline).toHaveLength(0)
        expect(planningSlugline).toHaveLength(0)
      })
    })

    describe('Title', () => {
      it('should have the same title in planning, assignment and flash', () => {
        const assignmentTitle = planning.document.meta.find((m: Block) => m.type === 'core/assignment')?.title

        expect(flash.document.title).toEqual(assignmentTitle)
        expect(planning.document.title).toEqual(assignmentTitle)
      })
    })

    describe('Misc', () => {
      it('should call sendStateless twice', () => {
        expect(sendStateless).toHaveBeenCalledTimes(2)
      })
    })
  })

  describe('Existing planning', () => {
    let flash: { document: Document }
    let planning: { document: Document }
    let sendStateless: ReturnType<typeof vi.fn>

    beforeAll(() => {
      ({ flash, planning, sendStateless } = runCreateFlash(true))
    })

    describe('Snapshots', () => {
      it('should match planning snapshot', () => {
        expect(planning).toMatchSnapshot()
      })

      it('should match flash snapshot', () => {
        expect(flash).toMatchSnapshot()
      })
    })

    describe('Section', () => {
      it('should only have one section block in each document', () => {
        const flashSection = flash.document.links.filter((l: Block) => l.type === 'core/section')
        const planningSection = planning.document.links.filter((l: Block) => l.type === 'core/section')

        expect(flashSection).toHaveLength(1)
        expect(planningSection).toHaveLength(1)
      })

      it('should have the same section in both planning and article', () => {
        const flashSection = flash.document.links.filter((l: Block) => l.type === 'core/section')
        const planningSection = planning.document.links.filter((l: Block) => l.type === 'core/section')

        expect(flashSection[0].title).toEqual(SECTION)
        expect(flashSection).toEqual(planningSection)
      })
    })

    describe('Slugline', () => {
      it('should not have a slugline block, neither in flash or planning', () => {
        const flashSlugline = flash.document.meta.filter((m: Block) => m.type === 'tt/slugline')
        const planningSlugline = planning.document.meta.filter((m: Block) => m.type === 'tt/slugline')

        expect(flashSlugline).toHaveLength(0)
        expect(planningSlugline).toHaveLength(0)
      })
    })

    describe('Title', () => {
      it('should have different title in planning and article', () => {
        const assignmentTitle = planning.document.meta.find((m: Block) => m.type === 'core/assignment')?.title

        expect(assignmentTitle).toEqual(ASSIGNMENT_TITLE)
        expect(flash.document.title).toEqual(assignmentTitle)
        expect(planning.document.title).toEqual(PLANNING_TITLE)
      })
    })

    describe('Misc', () => {
      it('should call sendStateless twice', () => {
        expect(sendStateless).toHaveBeenCalledTimes(2)
      })
    })
  })
})
