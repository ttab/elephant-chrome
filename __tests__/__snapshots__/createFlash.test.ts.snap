// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createFlash > Existing planning > Snapshots > should match flash snapshot 1`] = `
{
  "document": {
    "content": [
      {
        "content": [],
        "contenttype": "",
        "data": {
          "text": "Test flash",
        },
        "id": "uuid-38",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "heading-1",
        "sensitivity": "",
        "title": "",
        "type": "core/text",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
      {
        "content": [],
        "contenttype": "",
        "data": {
          "text": "",
        },
        "id": "uuid-39",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "",
        "type": "core/text",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
    "language": "sv-se",
    "links": [
      {
        "content": [],
        "contenttype": "",
        "data": {},
        "id": "",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "Inrikes",
        "type": "core/section",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
    "meta": [],
    "title": "Test flash",
    "type": "core/flash",
    "uri": "core://flash/uuid-37",
    "url": "",
    "uuid": "uuid-37",
  },
  "isMetaDocument": false,
  "mainDocument": "",
  "version": 0n,
}
`;

exports[`createFlash > Existing planning > Snapshots > should match planning snapshot 1`] = `
{
  "document": {
    "content": [],
    "language": "sv-se",
    "links": [
      {
        "content": [],
        "contenttype": "",
        "data": {},
        "id": "",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "Inrikes",
        "type": "core/section",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
    "meta": [
      {
        "content": [],
        "contenttype": "",
        "data": {
          "end_date": "2025-02-17",
          "public": "true",
          "start_date": "2025-02-17",
          "tentative": "false",
        },
        "id": "",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "",
        "type": "core/planning-item",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
      {
        "content": [],
        "contenttype": "",
        "data": {},
        "id": "",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "",
        "type": "core/newsvalue",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "4",
      },
      {
        "content": [],
        "contenttype": "",
        "data": {
          "end_date": "2025-02-17",
          "full_day": "false",
          "public": "true",
          "start": "2025-02-17T00:00:00.000Z",
          "start_date": "2025-02-17",
        },
        "id": "uuid-42",
        "links": [
          {
            "content": [],
            "contenttype": "",
            "data": {},
            "id": "",
            "links": [],
            "meta": [],
            "name": "",
            "rel": "deliverable",
            "role": "",
            "sensitivity": "",
            "title": "",
            "type": "core/flash",
            "uri": "",
            "url": "",
            "uuid": "uuid-37",
            "value": "",
          },
        ],
        "meta": [
          {
            "content": [],
            "contenttype": "",
            "data": {},
            "id": "",
            "links": [],
            "meta": [],
            "name": "",
            "rel": "",
            "role": "",
            "sensitivity": "",
            "title": "",
            "type": "core/assignment-type",
            "uri": "",
            "url": "",
            "uuid": "",
            "value": "flash",
          },
        ],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "Test flash",
        "type": "core/assignment",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
    "title": "Test planning",
    "type": "core/planning-item",
    "uri": "core://newscoverage/uuid-26",
    "url": "",
    "uuid": "uuid-26",
  },
  "isMetaDocument": false,
  "mainDocument": "",
  "version": 0n,
}
`;

exports[`createFlash > New planning > Snapshots > should match flash snapshot 1`] = `
{
  "document": {
    "content": [
      {
        "content": [],
        "contenttype": "",
        "data": {
          "text": "Test flash",
        },
        "id": "uuid-12",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "heading-1",
        "sensitivity": "",
        "title": "",
        "type": "core/text",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
      {
        "content": [],
        "contenttype": "",
        "data": {
          "text": "",
        },
        "id": "uuid-13",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "",
        "type": "core/text",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
    "language": "sv-se",
    "links": [
      {
        "content": [],
        "contenttype": "",
        "data": {},
        "id": "",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "section",
        "role": "",
        "sensitivity": "",
        "title": "Inrikes",
        "type": "core/section",
        "uri": "",
        "url": "",
        "uuid": "956636ed-2687-4bc3-a45a-7e09d98c6eeb",
        "value": "",
      },
    ],
    "meta": [],
    "title": "Test flash",
    "type": "core/flash",
    "uri": "core://flash/uuid-11",
    "url": "",
    "uuid": "uuid-11",
  },
  "isMetaDocument": false,
  "mainDocument": "",
  "version": 0n,
}
`;

exports[`createFlash > New planning > Snapshots > should match planning snapshot 1`] = `
{
  "document": {
    "content": [],
    "language": "sv-se",
    "links": [
      {
        "content": [],
        "contenttype": "",
        "data": {},
        "id": "",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "section",
        "role": "",
        "sensitivity": "",
        "title": "Inrikes",
        "type": "core/section",
        "uri": "",
        "url": "",
        "uuid": "956636ed-2687-4bc3-a45a-7e09d98c6eeb",
        "value": "",
      },
    ],
    "meta": [
      {
        "content": [],
        "contenttype": "",
        "data": {
          "end_date": "2025-02-17",
          "public": "true",
          "start_date": "2025-02-17",
          "tentative": "false",
        },
        "id": "",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "",
        "type": "core/planning-item",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
      {
        "content": [],
        "contenttype": "",
        "data": {},
        "id": "",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "",
        "type": "core/newsvalue",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "4",
      },
      {
        "content": [],
        "contenttype": "",
        "data": {
          "end_date": "2025-02-17",
          "full_day": "false",
          "public": "true",
          "start": "2025-02-17T00:00:00.000Z",
          "start_date": "2025-02-17",
        },
        "id": "uuid-16",
        "links": [
          {
            "content": [],
            "contenttype": "",
            "data": {},
            "id": "",
            "links": [],
            "meta": [],
            "name": "",
            "rel": "deliverable",
            "role": "",
            "sensitivity": "",
            "title": "",
            "type": "core/flash",
            "uri": "",
            "url": "",
            "uuid": "uuid-11",
            "value": "",
          },
        ],
        "meta": [
          {
            "content": [],
            "contenttype": "",
            "data": {},
            "id": "",
            "links": [],
            "meta": [],
            "name": "",
            "rel": "",
            "role": "",
            "sensitivity": "",
            "title": "",
            "type": "core/assignment-type",
            "uri": "",
            "url": "",
            "uuid": "",
            "value": "flash",
          },
        ],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "Test flash",
        "type": "core/assignment",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
    "title": "Test flash",
    "type": "core/planning-item",
    "uri": "core://newscoverage/uuid-1",
    "url": "",
    "uuid": "uuid-1",
  },
  "isMetaDocument": false,
  "mainDocument": "",
  "version": 0n,
}
`;
